\documentclass[11pt]{article}
\usepackage{longtable}
\usepackage{color}
\usepackage{tabu}
\usepackage{setspace}
\usepackage{pdflscape}
\usepackage{graphicx}
%\usepackage{subfigure}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{natbib}
\usepackage{fullpage}
%\bibliographystyle{plain}
\bibliographystyle{cbe}
\usepackage{algorithmic}
\usepackage[vlined,ruled]{algorithm2e}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage{url}
\usepackage[dvipsnames]{xcolor}
\usepackage[colorlinks=true, linkcolor=blue, citecolor=DarkOrchid, urlcolor=TealBlue ]{hyperref}
\usepackage[nottoc,numbib]{tocbibind}

\let \oriAlgorithm=\algorithm%
\renewcommand{\algorithm}[2]{%
	\oriAlgorithm%
	\SetAlgoLined%
	\DontPrintSemicolon%
	\SetKwComment{Comment}{$\triangleright\ $}{}%
	\TitleOfAlgo{ #1 }%
	\caption{ #2 }%
}%
\newcommand*{\Set}[2]{ #1 $\gets$ #2 \;}
\newcommand*{\SetC}[3]{ #1 $\gets$ #2 \Comment*[r]{#3}}
\newcommand*{\ForC}[3]{\For(\Comment*[f]{#2}){#1}{#3}}

\setcounter{tocdepth}{2} 

\begin{document}
	%\firstpage{1}
	
	\title{PhylogeneticGraph\\User Manual\\Version 0.1}
	%\subtitle{Program Documentation\\Version 0.1}
	\author{Ward C. Wheeler\\
		Division of Invertebrate Zoology,\\ American Museum of Natural History,\\ 200 Central Park West, New York, NY, 10024, USA;\\wheeler@amnh.org}
	
	
	\maketitle
%	\begin{center}
%		Running Title: {PhyGraph Version 0.1}
%	\end{center}
	\newpage
	
	\tableofcontents
	\newpage
	\section{Introduction}
	This is the initial version of documentation for the program PhylogeneticGraph (\texttt{PhyG}).  This program is designed to produce phylogenetic graphs from input data and graphs via heuristic searching of general phylogenetic graph space.
	
	 \texttt{PhyG} is a sucessor program to \texttt{POY} \citep{POY2,POY3,POY4,Varonetal2010,POY5, Wheeleretal2015} \url{https://github.com/wardwheeler/POY5}, containing much of its functionality but extended to broader classes of input data and phylogenetic graphs.
	The phylogenetic graph inputs and outputs of \texttt{PhyG} include not only trees, but other forms including forests and both
	soft and hard-wired networks.
	
	
	
	\section{Overview of program use}
	At present, \texttt{PhyG} is operated solely via command-line interface. 
	
	Commands are entered via a scriptfile that contains commands, which may themselves contain references to other script files.
	These commands specify input files, output files and formats, graph type and search parameters. 
	
	\bigskip
	\texttt{phyg script.pg}
	
	
	\section{QuickStart}
		\subsection{Requirements}
		\subsection{Obtaining and Installing \texttt{PhyG}}
			All source code, precompiled binaries, test data, and documentation are available from \url{https://githib.com/wardwheeler/PhyGraph}.

	\section{Input Data Formats}
		Any character names in input files are (for now) ignored and internal names are created
		by appending the character number in its file to the filename as in "fileName:0".
		Qualitative data, and prealigned data include their index in their input files and unaligned 
		data are treated as a single character.
		\subsection{fasta}
			Single character sequence input \citep{PearsonandLipman1988}.
		\subsection{fastc}
			Multicharacter sequence input.  \citep{WheelerandWashburn2019}.
		\subsection{\texttt{TNT}}
		The TNT \citep{Goloboffetal2008} format is accepted here for specification of qualitative, measurement, and prealigned molecular 
		sequence data.  \texttt{PhyG} does not parse all the diversity of options that can be specified in \texttt{TNT} input files.
		Do not support everything, interleave yes, cc code and costs yes, but one set of commands per line.
		Costs A$>$B A$/$B syntax no spaces. Ambiguities not allowed.  Must specify all transformations manually.
		')' sets to non-additive, if wantadditive then need to reset to additive after\\
		character state designations single characters\\
		continuous and other multicharacter states--can be read, abinguitess or ranges (unlike tnt for continuous) are 
		in squarebraskets with period as in [X.Y]\\
		- ans ? are always missing/inapplicable.  If DNA are not coded ACGT- => 01234 but left as lettes,
		then to include gaps as a 5th state, just include another character, such the letter 'O' that is not an IUPAC ambiguity code for DNA (or amino acids for that matter) as an additional state.  This can be used for matrix/Sankoff matrices as well.
		Amino acid sequences would be processed in the same way.  Including gaps as information requires
		an extra state (such as letter 'O').\\ 
		multi-character state designations (letters, numbers, etc) must be in their own ``block'' with spaces 
		between them.\\
		Continuous characters must be numbers only (float fine) and declares by cccode command as 
		``additive'', otherwise the number will be treated as non-additive character states. \\
		nonAdd polymorphisms  are [X.Y]--unless  additive '-' for range\\
		can't have '.' in multichar state (or single for that matter)\\
		Inherent ordering of DNA and AMino Acid codes is alphabetical (e.g. A,C,G,T.`-')
	
	\section{Input Graph Formats}
	Graphs may be input in the graphviz ``dot'' format \url{https://graphviz.org/}, Newick (as interpreted by Gary Olsen; \url{https://evolution.genetics.washington.edu/phylip/newick_doc.html}), Enhanced Newick \cite{Cardonaetal2008},
	and Forest Enhanced Newick (defined by \citealp{WheelerPhyloSuperGraphs}) formats.
	
	Quickly, Forest Enhanced Newick (FEN) is a format based on Enhanced Newick (ENewick) for forests of components, each
	of which is represented by an ENewick string.  The ENewick components are surrounded by `$<$' and '$>$'.
	As in $<$(A, (B,C)); (D,(E,F));$>$.  Groups may be shared among ENewick components.
	
	
	\section{Output Graph Formats}
	Graph outputs can be in either Graphviz `dot' or FEN formats.  Dot files can be visualized in a variety of ways 
	using Graphviz (e.g. dot, neanto, twopi) into pdf, jpg and a large variety of other formats. FEN outputs of 
	single trees (ie forest with a single component) are rendered as enewick.  Newick files can be visualized in a 
	large number of programs (e.g. FigTree; \url{http://tree.bio.ed.ac.uk/software/figtree/}, Dendroscope; 
	\url{https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/}). 	
	When FEN/Enewick files are output, leaf vertices are modified if they have indegree $>$ 1, creating a new node as parent to that leaf
	and redirecting the leaf's in-edges to that new node with a single edge connecting the new node to the leaf.  
	\bigskip
	\\Example dot command line: \\dot -Tpdf myDotFile.dot $>$ myDotFile.pdf\\
	
	Mulitple ``dot'' graphs can be output in a single file.  To create pdf and 
	other formats the commandline would be (these files are named and numbered automatically):\\\\dot -Tpdf -O myDotFile.dot\\
	
	For some reason on OSX the `pdf' option does not seem to work.  You can use `-Tps2' and that will generate a postscript file ($>$ blah.ps) that Preview can read and convert to pdf.
	
	\section{Command options}
	There are only a few program options that require specification.  There are defaults for all but input graphs.  
	Parameters are given with options in a range `a to b' (a-b) with any value in the interval, or alternates `a or b' (a|b). File options require a valid filename.
	For input graphs, wildcards are allowed (ie `*' and `?').  All commands are followed by a colon `:' before the option with no spaces.  Capitalization (for commands, but not filenames) is ignored.  Commands can be in any order (or entered from a file as stdin `$<$ filename).
	
	

	
	The program requires at least one input graph file and at least two input graphs (they could be in the same file).

	
	\section{Program Use}
	The program is invoked from the command-line as in:\\
	PhyGraph commandFile\\
	
	\section{Commands}
		\subsection{Command Structure}
		Commands and arguments are case insensitive with the exception of filename specifications, which are also 
		always in double quotes (\texttt{"fileName"}).  Commands are specified with mandatory parentheses \texttt{`()'} which may be empty or
		contain command arguments. Arguments may be preceded or followed by options separated by a colon \texttt{':'}.  Command arguments are not 
		order dependent.
		
		\bigskip
		\texttt{command(argument, option:argument, option:argument...)}
	
	\subsection{Build}
		\subsubsection{Syntax}
		\texttt{build(arg0, arg1:option, ...)}
		\subsubsection{Description}
		Builds initial graphs.  Most options (with the exception of the $O(n^3)$ distance based \texttt{dWag} option and character based \texttt{bWag})
		are $O(n^2)$.  Builds may be based on character methods or those based on distance methods.  Distance methods are considerably faster 
		(lower constant factor), but approximate in terms of character-based methods.  Refinement (branch swapping) can be specified
		for distance builds.  Given the large time burden, distance refinement is usually not time effective \citep{Wheeler2021}.
		\subsubsection{Arguments}
		\texttt{distance} Causes a pairwise distance matrix to be calculated ($O(n^2)$) and used as a basis for tree construction.
		
		\noindent \texttt{dWag} Performs distance Wagner build as in \citep{Farris1972} choosing the `best' taxon to add at each step, 
		yielding a single tree.  Time complexity of $O(n^3)$.
		
		\smallskip
		\noindent \texttt{rWag} Performs randoms addition sequence distance Wagner builds, yielding a multiple trees determined 
		by argument \texttt{replicates:m}.  Time complexity of $O(m \times n^2)$.
		
		\smallskip		
		\noindent \texttt{nj} Performs Neighbor-Joining build \citep{Saitou1987}, yielding a single tree.  Time complexity of $O(n^3)$.
		
		\smallskip		
		\noindent \texttt{wpgma} Performs Weighted Pair Group Method with Arithmetic Mean build \citep{SokalandMichener1958}, 
		yielding a single tree.  Time complexity of $O(n^2)$.
		
		\smallskip		
		\noindent \texttt{replicates:n}  Applies only to rWag.  Specifies the number of random addition sequences performed. 
		
		\smallskip		
		\noindent \texttt{best:n} Applies only to rWag.  Specifies the number of trees retained after rWag builds, selecting 
		the best trees in terms of distance cost.  The options can be used to reduce
		the number of trees retained for refinement or returned for further analysis.  
		
		\smallskip		
		\noindent \texttt{none} No refinement (OTU,SPR,TBR) is performed after distance builds.
		
		\smallskip		
		\noindent \texttt{otu} Specifies that OTU refinement \citep{Wheeler2021} is performed after distance builds.
		
		\smallskip		
		\noindent \texttt{spr} Specifies that SPR refinement \citep{Dayhoff1969} is performed after distance builds.

		\smallskip		
		\noindent \texttt{tbr} Specifies that TBR refinement \citep{Farris1988, swofford1990a} is performed after distance builds.

		\subsubsection{Defaults}
			\texttt{build(distance, dwag)}
		\subsubsection{Examples}
			\begin{itemize}
				\item{\texttt{build(distance, dWag, nj, wpgma)}\\Builds a single `best' addition sequence distance Wagner build, a Neighbor-Joining tree, and a WPGMA tree.}
				\item{\texttt{build(distance, rWag, replicates:1000, best:10)}\\Builds 1000 random addition sequence  Wagner builds and returns  
				10 of lowest distance cost.}
				\item{\texttt{build(distance, rWag, WPGMA, replicates:100, best:10, OTU)} \\Builds 100 random addition sequence  Wagner builds, a WPGMA tree, performs OTU swapping on the WPGMA and 10 of the lowest cost random addition sequence Wagner trees. 
			10 of lowest distance cost.}
			\end{itemize}
	\subsection{Read}
		\subsubsection{Syntax}
		\subsubsection{Description}
		Causes various information to input to the program.
		\subsubsection{Arguments}
		\subsubsection{Defaults}
		\subsubsection{Examples}

%	\subsection{Refine}
%		\subsubsection{Syntax}
%		\subsubsection{Description}
%		\subsubsection{Arguments}
%		\subsubsection{Defaults}
%		\subsubsection{Examples}
	
	\subsection{Rename}
		\subsubsection{Syntax}
		\subsubsection{Description}
			Command to rename taxa.
		\subsubsection{Arguments}
		\subsubsection{Defaults}
		\subsubsection{Examples}
	
	\subsection{Report}
		\subsubsection{Syntax}
		\subsubsection{Description}
			Outputs various information.
		\subsubsection{Arguments}
		\subsubsection{Defaults}
		\subsubsection{Examples}
	
	\subsection{Run}
		\subsubsection{Syntax}
		\subsubsection{Description}
			Used to execute a file containing commands.
		\subsubsection{Arguments}
		\subsubsection{Defaults}
		\subsubsection{Examples}
	
	\subsection{Select}
		\subsubsection{Syntax}
			\texttt{set(arg0:option, arg1:option, ...)}
		\subsubsection{Description}
			The command is used to set the method and number of graphs to be saved at any point.
			When multiple graphs are present, the \texttt{select} command will specify which of the graphs to
			keep for further analysis or reporting.
		\subsubsection{Arguments}
			\noindent \texttt{all} Keeps all graphs.
		
			\smallskip		
			\noindent	\noindent \texttt{best} Keeps graphs of best optimality value.
			
			\smallskip		
			\noindent \texttt{random} Keeps graphs chosen at random.
			
			\smallskip		
			\noindent \texttt{arg:n} Keeps up to \texttt{n} graphs determined by the argument.
		\subsubsection{Defaults}
			\texttt{select()} Keeps all unique graphs of best optimality value.
		\subsubsection{Examples}
				\begin{itemize}
					\item{\texttt{select(random:10)} Keeps up to 10 graphs, selected at random.}
					\item{\texttt{select(best:10)} Keeps up to 10 graphs of best optimality value.}
			\end{itemize}
	\subsection{Set}
		\subsubsection{Syntax}
				\texttt{set(arg0:option, arg1:option, ...)}
		\subsubsection{Description}
		The command specifies a number of global values and options include outgroup, graph type and optimality criterion.
		\subsubsection{Arguments}
		\subsubsection{Defaults}
		There are no default \texttt{set} arguments.  The default outgroup is the taxon whose name  is lexically first after
		renaming and taxon inclusion/exclusion. For this reason, it is best to specify an outgroup explicitly.
		The default optimality criterion is \texttt{parsimony} and the default graph type is \texttt{tree}.
		\subsubsection{Examples}
			\begin{itemize}
					\item{\texttt{set(optimality:parsimony)}}
			\end{itemize}
%	\subsection{Support}
%		\subsubsection{Syntax}
%		\subsubsection{Description}
%		\subsubsection{Arguments}
%		\subsubsection{Defaults}
%		\subsubsection{Examples}
%		
%	\subsection{Swap} 
%		\subsubsection{Syntax}
%		\subsubsection{Description}
%		\subsubsection{Arguments}
%		\subsubsection{Defaults}
%		\subsubsection{Examples}
%	
%	\subsection{Transform}
%		\subsubsection{Syntax}
%		\subsubsection{Description}
%		\subsubsection{Arguments}
%		\subsubsection{Defaults}
%		\subsubsection{Examples}
%	
	
	\section{Execution in Parallel}
	By default the program will execute using a single process core.  By specifying the options `+RTS -NX -RTS' where `X' is the number of processors offered to the program. These are specified after the program as in (for 4 parallel threads):\\
	\\
	PhyGraph +RTS -N4 -RTS other options...  \\
	
	\section*{Acknowledgments}
	The author would like to thank DARPA SIMPLEX N66001-15-C-4039, the  Robert J. Kleberg Jr. and Helen C. Kleberg foundation grant ``Mechanistic Analyses of Pancreatic Cancer Evolution'', and the American Museum of Natural History for financial support.  
	
	\newpage
	%\bibliography{big-refs-3.bib}
	\bibliography{/users/ward/Dropbox/Work_stuff/manus/big-refs-3.bib}
	%\bibliography{/home/ward/Dropbox/Work_stuff/manus/big-refs-3.bib}
\end{document}