\documentclass[11pt]{article}
\usepackage{longtable}
\usepackage{color}
\usepackage{tabu}
\usepackage{setspace}
\usepackage{pdflscape}
\usepackage{graphicx}
\usepackage {float}
%\usepackage{subfigure}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{natbib}
\usepackage{fullpage}
\bibliographystyle{plain}
%\bibliographystyle{cbe}
\usepackage{algorithmic}
\usepackage[vlined,ruled]{algorithm2e}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage[T1]{fontenc}
\usepackage{url}
 
\usepackage[dvipsnames]{xcolor}
\usepackage{color, soul}
\usepackage[colorlinks=true, linkcolor=blue, citecolor=DarkOrchid, urlcolor=TealBlue ]{hyperref}
%\usepackage[nottoc,numbib]{tocbibind}
\usepackage{tocloft}


\setlength\itemindent{0.25cm}

\newcommand{\phyg}{\texttt{PhyG} }
\newcommand{\BigO}[1]{\ensuremath{\mathcal{O}\left(\,#1\,\right)}\xspace}

\title{PhyG 0.3 Tutorials}

\author{Louise M. Crowley}
\makeindex
\begin{document}
\maketitle

\section{\phyg Tutorials}

These tutorials provide guidance for performing network analyses using the 
phylogenetic program \texttt{PhyG}. In addition to illustrating the vertical transfer 
of information between ancestor-descendant lineages, phylogenetic networks 
can convey information about reticulation events such as horizontal gene transfer, 
hybridization and introgression between lineages. There are two main types of 
networks---`softwired' and `hardwired'. When displayed, these networks main 
appear similar, however, they represent different interpretations of the meaning 
of phylogenetic edges.

A softwired network is a summary of a set of individual `display' trees or `tree-based 
networks', which have been generated by removing edges from the network.
In softwired networks, individual characters have single parents, as opposed 
to those in hardwired networks, where characters can have multiple parents 
\citep{KannanandWheeler2012}.

The user is advised to work through Tutorial 0.1 prior to attempting this tutorial, 
as this document provides instructions on obtaining and installing \texttt{PhyG}, 
as well as making and executing scripts. As in previous tutorials, each tutorial 
contains a \phyg script that includes detailed commentaries explaining the 
rationale behind each step of the analysis. The values of arguments herein have 
been chosen such that the analysis can complete within the timeframe of this 
session. Therefore, the values used here should not be taken to be optimal 
parameters. 

These tutorials use sample datasets that can also be downloaded from the 
\texttt{PhyG} \href{https://github.com/amnh/PhyGraph}{GitHub} website. Move 
these data files to a directory on your Desktop called \textbf{phygfiles}. The 
minimally required items to run the tutorial analyses are the \phyg application 
and sample data files. Running these analyses requires some familiarity with 
the \phyg command structure, to which a complete guide can be found 
\href{https://github.com/amnh/PhyGraph}{here}.

%-------------------------------------------------------------------------------------------------------
\subsection{Making a script and inspecting the data}
\label{subsec:networkscript}

In this tutorial, you will generate the initial script for a softwired network analysis 
and inspect the inputted data.

\begin{enumerate}

\item Open your text editor of choice and type the following:

	\begin{quote}	
	-\/-building networks for softwired tests using flu datasets\\
	set(seed:1634561640)\\
	set(outgroup:"1466\_H7N2\_Avian\_chicken\_pa\_1490921\_2002")\\
	set(graphtype:softwired)\\
	set(graphfactor:nopenalty)\\ 
	read(nucleotide:"flu*.fas*", tcm:(2,1))\\
	report("flu\_net1\_data.csv", data, overwrite)\\
	report("flu\_net1\_cr.csv", crossrefs, overwrite)
	\end{quote}

\item The script begins with a comment that describes the purpose of this 
analysis. Recall that comments are prepended with `-{}-' and can span multiple 
lines, provided each line begins with `-{}-'. \phyg will ignore any commented lines.\\

In the next four lines, we change the settings of \texttt{PhyG}. All \texttt{set}
commands are executed at the start of a run, irrespective of where they appear 
in the script. 

\item First, we \texttt{set} the seed for the random number generator to the 
integer value 1634561640. By setting this value, we are guaranteed to reproduce 
a given search trajectory each time the script is run.

\item Next, the outgroup for the analysis is \texttt{set} to 
\emph{1466\_H7N2\_Avian\_chicken\_pa\_1490921\_2002}. If the outgroup is not 
\texttt{set}, the default outgroup is the taxon whose name is lexically first after any 
renaming of taxa, and/or if taxa were specified by using the arguments \texttt{include}
or \texttt{exclude}. Only a single taxon can be set as the outgroup of the analysis. 
Recall that taxon names cannot have spaces, otherwise the names can be incorrectly 
interpreted by the program.

\item We next \texttt{set} the \texttt{graphtype} of this analysis. \phyg allows for 
the input, analysis of and output of a broader class of phylogenetic graphs. These
include trees, forests and both softwired and hardwired networks. The current choices 
are \texttt{tree} (the default), \texttt{hardwired} and \texttt{softwired}. We \texttt{set} 
the \texttt{graphtype} to \texttt{softwired}.

\item When conducting a network analysis, a penalty can be ascribed, so that 
softwired phylogenetic networks can compete equally with phylogenetic trees on 
a parsimony optimality basis. A network penalty takes into account the change in 
cost as edges are added to the tree. Current choices include \texttt{nopenalty}, 
\texttt{w15} and \texttt{w23}. In general, \texttt{w23} is a more severe penalty than
\texttt{w15}. For the generation of an initial network for further analysis, we 
\texttt{set} the \texttt{graphfactor} to \texttt{nopenalty}. Note: assigning \texttt{nopenalty}
is fine for the generation of graphs for further refinement, however it is unlikely 
to be a reasonable penalty for `real' analyses.

\item Change to the directory where the downloaded data files are located by using the 
\texttt{cd} command, as in:
		
	\begin{quote}
	cd ~/Desktop/phygfiles
	\end{quote}

By typing \texttt{ls} you will see that this directory contains twelve files in fasta format.
A primary requirement of this type of analysis is to have a minimum of two blocks of 
data. The command \texttt{read} imports our data files. Rather than type in the name 
of each of these files in our script, we will use wildcards (*) to capture all these files: 
        
        \begin{quote}
	read(nucleotide:"flu*.fas*", tcm:(2,1)\\
	\end{quote}

Filenames must include the suffix (e.g. .fas, .fasta, .fastc, .ss, .tre). Note: in this case, 
the wildcards capture files ending in .fasta and .fas. Failure to include these suffices 
will result in the error "File(s) not found in `read' command". The filename must match 
\textit{exactly}, including capitalization. We indicate that the files contain IUPAC 
\texttt{nucleotide} sequence data in fasta format. The argument \texttt{tcm} refers to 
the transformation cost matrix. The first integer specifies the substitution cost and the 
second integer value defines the indel cost. By default, the cost of both are set to 1.

\item Having read in our data, it is advisable to verify that the files were properly 
parsed. We \texttt{report} a \texttt{crossrefs} and \texttt{data} files, which allows 
us to output information concerning the characters and terminals in our data files. 

\item Save this file with the name \textbf{flu\_net1.pg} in a directory \texttt{phygfiles} 
located on your Desktop.

\item Run the script by typing the following:

	\begin{quote}
  	phyg flu\_net1.pg
	\end{quote}

Notice the output appear in the \textit{Terminal} window.

\item Examine the reported files \textbf{"flu\_net1\_data.csv"} and 
\textbf{"flu\_net1\_cr.csv"}. The \texttt{data} file summarizes information relating 
to the input data (number of terminals, number of input files, number of character 
blocks and the total number of characters). The \texttt{crossrefs} file provides a 
comprehensive visual overview of the completeness of the data. Together these 
files indicate that data has been input for nine terminal taxa, across 12 data 
blocks. The \texttt{crossrefs} file highlights that data in two of the files is missing 
for three taxa.

\end{enumerate}

%-------------------------------------------------------------------------------------------------------
\subsection{Generating a network from trees}
\label{subsec:makinganetwork}

The first step in the analysis of networks is to generate one. Having imported and 
inspected our data, we are now ready to build the initial graphs. In this tutorial, you 
will learn how to generate a network from trees using \texttt{PhyG}. 

\begin{enumerate}

\item Modify the script to include \texttt{build(distance, rdwag, block, graph, eun, 
replicates: 1000)}. By specifying \texttt{block}, \phyg performs independent builds 
for each ``block'' of data. If this option is not specified, the builds are performed 
combining all the data. This command causes a pairwise distance matrix to be 
calculated ($\BigO n^2$) that is subsequently used as a basis for distance tree 
construction. \texttt{distance} trees are constructed on the \texttt{block}s of data 
by performing random addition sequence distance Wagner builds (\texttt{rdWag}), 
yielding multiple trees (\texttt{n}), as determined by the argument \texttt{replicates}. 
The resulting trees or graphs are reconciled using the \texttt{eun} argument, turning 
them into a network. \texttt{eun} reconciles block trees into a Edge-Union-Network 
\citep{MiyagiandWheeler2019, Wheeler2022}.

\item We will want to examine the resulting trees. Trees are not reported as 
output in the \textit{Terminal} window and must be directed to a file. Modify the 
script to output tree files with \texttt{report("flu\_net1.tre", graphs, newick, 
overwrite)} and \texttt{report("flu\_net1\_gv.tre", dotpdf, graphs, overwrite)}.

	\begin{quote}	
	-\/-building networks for softwired tests using flu datasets\\
	set(seed:1634561640)\\
	set(outgroup:"1466\_H7N2\_Avian\_chicken\_pa\_1490921\_2002")\\
	set(graphtype:softwired)\\
	set(graphfactor:nopenalty)\\ 
	read(nucleotide:"flu*.fas*", tcm:(2,1))\\
	build(distance, rdwag, block, graph, graph, eun, replicates:1000)\\
	report("flu\_net1\_data.csv", data, overwrite)\\
	report("flu\_net1\_cr.csv", crossrefs, overwrite)\\
	report("flu\_net1.tre", graphs, newick, overwrite)\\
	report("flu\_net1\_gv.tre", dotpdf, graphs, overwrite)
	\end{quote}
	
\item Run the script.

\item Let's examine the reported tree files. Because the file \textbf{"flu\_net1.tre"} 
represents a phylogenetic network (as opposed to tree), this file can not be viewed 
using tree viewing programs such as \textit{FigTree} or \textit{TreeView}. Using 
your preferred text editor, open this file (Figure \ref{tre1}). This looks like a standard 
newick tree file, in parenthetical notation, however, notice the addition of \hl{four} 
\textbf{\#}'s, which represent reticulation events. The values associated with 
the taxon names and HTUs are the branch lengths. The cost of the tree(s) can 
be found in square brackets, at the end of each tree. In this analysis, \phyg returned 
a single network  with a cost of 13930.
 
\begin{figure}[H]
\centering
\includegraphics[width=\textwidth]{tre1.png}
\caption{\textbf{"flu\_net1.tre"}.}
\label{tre1}
\end{figure}

\item  The command \texttt{report("filename.tre", dotpdf, graphs, overwrite)} will 
produce a file that can be read in \textit{Adobe Illustrator}, \textit{Apple Preview} 
or any vectorial image editor program. Notice that two files were outputted from 
using this command. \phyg has output an eps (on OSX) or pdf (on linux) file that 
can be viewed in a vector graphics program and a dot file, which can be viewed 
(and modified) in \textit{Graphviz}. Note: in order to output pdf files the application 
\textit{dot} must be installed from the \href{https://graphviz.org/download/}{Graphviz} 
website. \textit{dot} is a graph description language and Graphviz an open-source 
graph visualization software. This program is well suited to representing graphs 
and networks. Open the reported file \textbf{"flu\_net1\_gv.tre.eps"} in your preferred
visualization program (Figure \ref{eps1}). The values associated with the taxon 
names and HTUs are the branch lengths. Examining this file, we can see \hl{two} 
networks (one leading into\hl{ HTU15 and HTU16}).

\begin{figure}[H]
\centering
\includegraphics[width=\textwidth]{eps1.png}
\caption{\textbf{"flu\_net1\_gv.tre.eps"}.}
\label{eps1}
\end{figure}

\end{enumerate}
%-------------------------------------------------------------------------------------------------------
\subsection{Refining a network: adding network edges}
\label{subsec:netadd}

Now that we have a network to work with, we can do some refinement operations 
on the network edges. In the next four tutorials we will learn how to perform 
manipulation of network edges to existing network graphs. This tutorial will 
focus on adding network edges.

\begin {enumerate}

\item Rather than modify the previous script (possible to do with comments and 
additional text), we will generate a new script to perform the refinements of our
first network.

\item Open your text editor of choice and type the following:

	\begin{quote}	
	-\/-Refinements of networks using flu data sets using netadd\\
	set(seed:1634561640)\\
	set(outgroup:"1466\_H7N2\_Avian\_chicken\_pa\_1490921\_2002")\\
	set(graphtype:softwired)\\
	set(graphfactor:w15)\\ 
	read(nucleotide:"flu*.fas*", tcm:(2,1))\\
	read(newick:"flu\_net1.tre")\\
	refine(netadd, maxnetedges:5, atrandom, steepest)\\
	report("flu\_net2.tre", graphs, newick, overwrite)\\
	report("flu\_net2\_gv.tre", dotpdf, graphs, overwrite)
	\end{quote}

\item The script begins with a comment that describes the purpose of this 
analysis.

\item The next three lines are identical to that of \textbf{flu\_net1.pg}, so we 
will not comment about them any further. 

\item Unlike the previous script, here, we \texttt{set} the \texttt{graphfactor} to 
\texttt{w15}. This penalty involves the calculation of the most Parsimony 
display tree. Because this penalty has a higher time complexity to calculate 
the total network costs (than \texttt{w23} and \texttt{nopenalty}), this may take 
significantly more time for larger datasets. 

\item In addition to importing the nucleotide sequence data files, we also
read in the \texttt{newick} tree file \textbf{"flu\_net2.tre"} generated from the 
previous tutorial. Reading in tree or graph files, rather than building them from 
scratch, can be a useful starting point and can speed up analyses.

\item Having read in our data and graph files, we can now perform refinements 
of the network. The command \texttt{refine} performs edit operations on network 
edges. The network specific refinement operations include \texttt{netadd}, 
\texttt{netdel}, \texttt{netadddel} and \texttt{netmove}. These refinements are 
only applied to network edges, hence they can only be applied to softwired and 
hardwired graphs (as opposed to trees). \\

The argument \texttt{netadd} adds network edges to input graphs at all possible
positions until no better cost graph is found. Though no need to specify here, 
the default arguments of \texttt{refine} are \texttt{atrandom} and \texttt{steepest}.
\texttt{atrandom} randomizes the evaluation of the networks and the order by 
which they are generated. \texttt{steepest} specifies that if a better network is 
found during the network refinement operations then the refinements will switch 
greedily to this graph and abandon the previous network. The argument  
\texttt{maxnetedges:n} only applies to \texttt{netadd}. This specifies that network
edges can only be added to networks with the number of edges less that the integer 
specified in \texttt{maxnetedges:n}.

\item Save this file with the name \textbf{flu\_net2.pg} in the same directory as the 
data files and previously reported tree files.

\item Run the script.

\item \hl{Examine the output in the \textit{Terminal} window. Here we can see that 
\phyg performed 3 rounds of netadd.}\\

Let's examine the reported tree files. 

\item Opening the file \textbf{"flu\_net2.tre"} in your preferred text editor. This 
newick file has ten \textbf{\#}'s, representing the reticulation events (Figure 
\ref{tre2}). Comparing this tree to \textbf{flu\_net1.tre} (our input tree), we can 
see that the addition of two additional network edges, reduced the cost of the 
network from 13930 to 13528. 

\begin{figure}[H]
\centering
\includegraphics[width=\textwidth]{tre2.png}
\caption{\textbf{"flu\_net2.tre"}.}
\label{tre2}
\end{figure}

\item  Open the reported file \textbf{"flu\_net2\_gv.tre.eps"} in your preferred
visualization program (Figure \ref{eps2}). The values associated with the taxon 
names and HTUs are the branch lengths. Examining this file, we can see five 
networks (one leading into HTU19, HTU24, HTU26, HTU22 and HTU16).

\begin{figure}[H]
\centering
\includegraphics[width=\textwidth]{eps2.png}
\caption{\textbf{"flu\_net2\_gv.tre.eps"}.}
\label{eps2}
\end{figure}

\end{enumerate}

%-------------------------------------------------------------------------------------------------------
\subsection{Refining a network: deleting network edges}
\label{subsec:netdel}

This tutorial will focus on deleting network edges.

\begin {enumerate}

\item Open your text editor of choice and type the following:

	\begin{quote}	
	-\/-Refinements of networks using flu data sets using netdel\\
	set(seed:1634561640)\\
	set(outgroup:"1466\_H7N2\_Avian\_chicken\_pa\_1490921\_2002")\\
	set(graphtype:softwired)\\
	set(graphfactor:w15)\\ 
	read(nucleotide:"flu*.fas*", tcm:(2,1))\\
	read(newick:"flu\_net1.tre")\\
	refine(netdel, atrandom, steepest)\\
	report("flu\_net3.tre", graphs, newick, overwrite)\\
	report("flu\_net3\_gv.tre", dotpdf, graphs, overwrite)
	\end{quote}

\item The script begins with a comment that describes the purpose of this 
analysis.

\item The next six lines are identical to that of \textbf{flu\_net1.pg}, so we 
will not comment about them any further. 

\end{enumerate}
%-------------------------------------------------------------------------------------------------------
\subsection{Refining a network: adding and deleting network edges}
\label{subsec:netdel}

This tutorial will focus on iteratively adding and deleting network edges.

\begin {enumerate}

\item Open your text editor of choice and type the following:

	\begin{quote}	
	-\/-Refinements of networks using flu data sets using netadddel\\
	set(seed:1634561640)\\
	set(outgroup:"1466\_H7N2\_Avian\_chicken\_pa\_1490921\_2002")\\
	set(graphtype:softwired)\\
	set(graphfactor:w15)\\ 
	read(nucleotide:"flu*.fas*", tcm:(2,1))\\
	read(newick:"flu\_net1.tre")\\
	refine(netadddel, atrandom, steepest)\\
	report("flu\_net4.tre", graphs, newick, overwrite)\\
	report("flu\_net4\_gv.tre", dotpdf, graphs, overwrite)
	\end{quote}

\end{enumerate}
%-------------------------------------------------------------------------------------------------------
\subsection{Refining a network: moving network edges}
\label{subsec:netdel}

This tutorial will focus on moving network edges.

\begin {enumerate}

\item Open your text editor of choice and type the following:

	\begin{quote}	
	-\/-Refinements of networks using flu data sets using netmove\\
	set(seed:1634561640)\\
	set(outgroup:"1466\_H7N2\_Avian\_chicken\_pa\_1490921\_2002")\\
	set(graphtype:softwired)\\
	set(graphfactor:w15)\\ 
	read(nucleotide:"flu*.fas*", tcm:(2,1))\\
	read(newick:"flu\_net1.tre")\\
	refine(netmove, atrandom, steepest)\\
	report("flu\_net5.tre", graphs, newick, overwrite)\\
	report("flu\_net5\_gv.tre", dotpdf, graphs, overwrite)
	\end{quote}

\end{enumerate}
%-------------------------------------------------------------------------------------------------------

%\printindex
\bibliography{/Users/louise/DropboxAMNH/big-refs-3.bib}
\end{document}
