\documentclass[11pt]{book}
\usepackage{longtable}
\usepackage{color}
\usepackage{tabu}
\usepackage{setspace}
\usepackage{pdflscape}
\usepackage{graphicx}
\usepackage {float}
%\usepackage{subfigure}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{natbib}
\usepackage{fullpage}
\bibliographystyle{plain}
%\bibliographystyle{cbe}
\usepackage{algorithmic}
\usepackage[vlined,ruled]{algorithm2e}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage[T1]{fontenc}
\usepackage{url}
 
\usepackage[dvipsnames]{xcolor}
\usepackage{color, soul}
\usepackage[colorlinks=true, linkcolor=blue, citecolor=DarkOrchid, urlcolor=TealBlue ]{hyperref}
%\usepackage[nottoc,numbib]{tocbibind}
\usepackage{tocloft}

\setlength\itemindent{1cm}

\newcommand{\phyg}{\texttt{PhyG} }
\newenvironment{phygdescription}{\subsubsection{Description}}{}
\newenvironment{example}{\subsubsection{Examples} \begin{itemize}}{\end{itemize}}
\newenvironment{argument}{\subsection{Arguments}\begin{itemize}}{\end{itemize}}
%\item [ ]

% We define a command environment for new command definitions, and store
% whatever command we are dealing with in the @commandname macro. Inside a
% command we can specify the defaults, the syntax, and the arguments to be used.
  \newenvironment{command}[2]{
	\def\tmpa{}
	\def\tmpb{#2}
	\def\@commandname{#1}
	\subsection{#1}\index{general}{#1}
	\ifx\tmpa\tmpb 
	\label{comm:#1} 
	\else 
	\label{comm:#2} 
	\fi}
{}

% Syntax definition. We use the name of the command as stored in @commandname
\newcommand{\syntax}{\subsubsection{Syntax} \@commandname} 
\newcommand{\atsymbol}{@}

\begin{document}
	%\firstpage{1}
	
	\title{PhylogeneticGraph\\User Manual\\Version 0.1}
	
	\maketitle
	
	\newpage

	 \begin{center}
		\includegraphics[width=\textwidth]{AMNHLogo.jpg}
	\end{center}

	\vspace*{5.50cm}	
	\begin{flushleft}
		\textbf {Program and Documentation} \\ Ward C. Wheeler \\
		\vspace*{0.50cm}
		\textbf {Program} \\ Alex Washburn \\
		\vspace*{0.50cm}
		\textbf{Documentation} \\ Louise M. Crowley
	\end{flushleft}
	

	\vspace*{5.50cm}
	
	\begin{flushleft}
		\small
		{\it Louise M. Crowley, Alex Washburn, Ward C. Wheeler} \\
		
		Division of Invertebrate Zoology, American Museum of Natural History, New York, NY, U.S.A.\\
		\smallskip
		The American Museum of Natural History\\
		\copyright 2022 by The American Museum of Natural History, \\
		All rights reserved. Published 2022.
		
		%	\vspace*{0.25cm}
		%	\emph{W. C. Wheeler.} 2022. \texttt{PHYG} 1..0. New York, 
		%	American Museum of Natural History. Documentation by W.C. Wheeler and L. M. Crowley. 
		%	
		\vspace*{0.25cm}
		
		Available online at \url{https://github.com/amnh/PhyGraph}
		
		Comments or queries relating to the documentation should be sent to \href{mailto:wheeler@amnh.org}
		{wheeler@amnh.org} or \href{mailto:crowley@amnh.org}{crowley@amnh.org}
	\end{flushleft}
	
	\tableofcontents

\chapter{What is PhyG?}

\section{Introduction}
	PhylogeneticGraph (\texttt{PhyG}) is a multi-platform program designed to produce phylogenetic 
	graphs from input data and graphs via heuristic searching of general phylogenetic graph 
	space. \texttt{PhyG} is the successor of \href{https://github.com/wardwheeler/POY5}{\textbf{POY}}
	\citep{POY2,POY3,POY4,Varonetal2010,POY5, Wheeleretal2015}, containing much of its 
	functionality, including the optimization of \textit{unaligned} sequences, and the ability to implement 
	search strategies such as random addition sequence, swapping, and tree fusing. As in {\textbf{POY}, 
	\phyg can generate outputs in the form of implied alignments and graphical representations of 
	cladograms and graphs. What sets \phyg apart from {\textbf{POY}, and other phylogenetic 
	analysis programs, is the extension to broader classes of input data and phylogenetic graphs. 
	The phylogenetic graph inputs and outputs of \texttt{PhyG} include trees, as well as other forms 
	including forests and both soft- and hard-wired networks.
		
	This is the initial version of documentation for the program.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%QUICKSTART
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%		
\section{Quick Start}
	
	\subsection{Requirements: software and hardware}
		\phyg is an open-source program that can be compiled for Mac OSX and Linux. Some 
		utility programs (such as TextEdit for Mac, or Nano for Linux) can help in preparing 
		\phyg scripts and formatting data files, while others (such as Adobe Acrobat and TreeView 
		\citep{page1996}) can facilitate viewing the outputted graphs and trees.
		
		\phyg runs on a variety of computers, including desktops, laptops and cluster computers.
		By default, \phyg is a multi-threaded application and will use the available resources of 
		the computer during the analysis (see Execution in Parallel \ref{subsec:parallel}). 
		
	\subsection{Obtaining and Installing \phyg}
		\phyg source code, precompiled binaries, test data, and documentation in pdf format, 
		as well as tutorials, are available from the \phyg \href{https://github.com/amnh/PhyGraph}{GitHub} 
		website.

	\subsubsection{Installing from the binaries}
		Download the \phyg binary from the \href{https://github.com/amnh/PhyGraph}{GitHub} 
		website. Binaries are available for Mac OSX computers with either Intel or M1 processors, 
		and Linux machines (see information relating to Windows machine below).
		The user can go directly to the website and click on the appropriate link 
		for the binary. On most systems this will download to either your Desktop or Downloads folder. \\
		
		Alternatively, open a \textit{Terminal} window (located in your Applications folder) and type 
		the following for either the Mac Intel, Mac M1 or Linux binary:
		
		\begin {quote}
		curl -LJ --output phyg https://github.com/amnh/PhyGraph/blob/main/bin/OSX/phyg-Intel?raw=true
		\end{quote}		
		
		\noindent or 
		
		\begin {quote}
		curl -LJ --output phyg https://github.com/amnh/PhyGraph/blob/main/bin/OSX/phyg-M1?raw=true
		\end{quote}	
		
		\noindent or 
		
		\begin {quote}
		curl -LJ --output phyg https://github.com/amnh/PhyGraph/blob/main/bin/linux/phyg?
		raw=true
		\end{quote}
		
		\noindent The binary should either be moved into your \$PATH or referred to its 
		absolute when executing a script.\\
		
		For those users with Windows machines, a Windows Subsystem for Linux 
		(WSL) can be installed. This system will allow you to run the Linux binary directly 
		on your machine, without having to install a virtual machine or dual-boot setup. 
		The WSL, along with directions for installation, can be found 
		\href{https://learn.microsoft.com/en-us/windows/wsl/}{here}.
	
	\subsubsection{Compiling from the source}
		For the majority of users, downloading the binaries will suffice. Should the user prefer to 
		compile \phyg directly from the source, the source code can be downloaded 
		from the \href{https://github.com/amnh/PhyGraph}{GitHub} website. \phyg is largely 
		written in Haskell. In order to compile \phyg from the source, the user must install Cabal, 
		a command-line program for downloading and building software written in Haskell (ghc).  
		Information on its installation can be found  
		\href{https://www.schoolofhaskell.com/user/simonmichael/how-to-cabal-install}{here}.
		To install an optimized version of \texttt{PhyG}, open a \textit{Terminal} and run the 
		following:
		
		\begin{quote}
		cabal install PhyGraph:phyg --project-file=cfg/cabal.project.release
		\end{quote}
		
		%\hl{Explain this a bit better?}
		
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%OVERVIEW OF THE PROGRAM
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	 		
\section{Overview of program use}
	At present, \phyg is operated solely via command-line in a \textit{Terminal} window
	and cannot be performed interactively. Commands are entered via a script file 
	containing commands that specify input files, output files and formats, graph type 
	and search parameters.
		
	\subsection{Executing Scripts}
		The program is invoked from the command-line as in:\\
		
		\begin{quote}
		phyg commandFile
		\end{quote}

		\noindent For example, typing the following in a \textit{Terminal} window will invoke 
		\phyg to run the script \texttt{mol.pg}, which is located in the Desktop folder 
		\texttt{phygfiles}:\\
		
		\begin{quote}
  		phyg Users/Ward/Desktop/phygfiles/mol.pg
		\end{quote}

		\noindent 
		This is the equivalent of typing the following from any location on your computer:\\
		
		\begin{quote}
   		cd ("Users/Ward/Desktop/phygfiles")
		\end{quote}
			
		\subsection{Creating and running \phyg scripts}
		A script is a simple text file containing a list of commands to be performed. Running 
		analyses using scripts allows for the entire analysis to proceed from the beginning to 
		the end with one click of a button and may produce faster results as \phyg automatically 
		optimizes the workflow of the analysis by considering the functional relationships 
		among various tasks and efficiently distributing the jobs and resources (such as memory 
		and multiple processors). This script 	can also include references to other script files 
		(Figure \ref{firstscript}).
		
		Scripts can be created using any conventional text editor such as TextEdit, TextWrangler, 
		BBEdit, or Nano. Comments that describe the contents of the file and provide other useful 
		annotations can be included. Comments are prepended with `-{}-' and can span multiple 
		lines, provided each line begins with `-{}-'. 

		\begin{figure}[H]
		\centering
		\includegraphics[width=\textwidth]{First_run.jpg}
		\caption{\phyg scripts. The headers in the scripts are comments, leading with `-{}-', which is 
		ignored by \phyg. This first script ``First\_script.pg'' includes a reference to the second script
		``Chel\_files.txt'', which includes a group of data files to be read by the program.}
		\label{firstscript}
		\end{figure}

% Optimize memory consumption--keep low number of graphs in initial searches, later keep 
% a larger number to get others 
%		\hl{(see email from WW 04-21-22)}
%		\hl{where should I discuss this?}
	
	\subsection{Execution in Parallel}
		\label{subsec:parallel}
		\phyg is a multi-threading application and will, by default, spawn as many jobs as there 
		are cores in the machine and can be more efficient. Should the user wish to limit or 
		specify the number of processors used by \phyg this can be achieved by including the 
		options `\textbf{+RTS -NX -RTS}', where `\textbf{X}' is the number of processors offered 
		to the program, when executing the script. Should the user wish to use a single processor, 
		this can be specified by typing:

		\begin{quote}
		phyg fileName +RTS -N1 
		\end{quote}		
		%possible to include and other options

		
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%FORMATS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Input Data Formats} 
	Any character names in input files are (for now) ignored and internal names are created
	by appending the character number in its file to the filename as in "fileName:0".
	Qualitative data, and prealigned data include their index in their input files and unaligned 
	data are treated as a single character.
		
	\subsection{fasta}
		Single character sequence input \citep{PearsonandLipman1988}.
		
	\subsection{fastc}
		Multicharacter sequence input. \citep{WheelerandWashburn2019}.
		
	\subsection{\texttt{TNT}}
		The TNT \citep{Goloboffetal2008} format is accepted here for specification of qualitative,
		measurement, and prealigned molecular sequence data. \phyg does not parse all the
		diversity of options that can be specified in \texttt{TNT} input files.\\
		
		%\hl{[Notes to be fleshed out...]}\\
%		The majority of character scopes of TNT files are supported in \phyg. These include interleaved files,
%		 cc code and costs yes, but one set of commands per line.\\
%		Costs A$>$B A$/$B syntax no spaces. \\
%		Ambiguities not allowed. \\
%		Must specify all transformations manually.
%		')' sets to non-additive, if want additive then need to reset to additive after\\
%		character state designations single characters\\
%		continuous and other multicharacter states--can be read, ambiguities or ranges (unlike \texttt{TNT} 
%		for continuous) are 
%		in squarebrakets with period as in [X.Y]\\
%		- ans ? are always missing/inapplicable. \\
%		If DNA are not coded ACGT- => 01234 but left as letters, then to include gaps as a 5th state, just 
%		include another character, such the letter 'O' that is not an IUPAC 
%		ambiguity code for DNA (or amino acids for that matter) as an additional state. This can be used for 
%		matrix/Sankoff matrices as well.\\
%		Amino acid sequences would be processed in the same way. Including gaps as information requires
%		an extra state (such as letter 'O').\\ 
%		multi-character state designations (letters, numbers, etc) must be in their own ``block'' with spaces 
%		between them.\\
%		Continuous characters must be numbers only (float fine) and declared as ``additive'' by cccode 
%		command, otherwise the number will be treated as non-additive character states. \\
%		nonAdd polymorphisms are [X.Y]--unless additive '-' for range\\
%		can't have '.' in multichar state (or single for that matter)\\
%		The inherent ordering of DNA and amino acid codes is alphabetical (e.g. A, C, G, T and `-').
%	
\section{Input Graph Formats}
	Graphs can be input in the graphviz \href{https://graphviz.org/}{``dot''} format Newick (as 		
	interpreted by Gary Olsen, linked \href{https://evolution.genetics.washington.edu/phylip/newick_doc.html}
	{here}), Enhanced Newick \cite{Cardonaetal2008}, and Forest Enhanced Newick (defined by 
	\citealp{Wheeler2022}) formats.
	Forest Enhanced Newick (FEN) is a format based on Enhanced Newick (ENewick) for 
	forests of components, each of which is represented by an ENewick string. The ENewick 
	components are surrounded by `$<$' and '$>$'. As in $<$(A, (B,C)); (D,(E,F));$>$. 
	Groups may be shared among ENewick components.
	
\section{Output Graph Formats}
	Graph outputs can be in either Graphviz `dot' or FEN formats. Dot files can be visualized 
	in a variety of ways using Graphviz (e.g. dot, neanto, twopi) into pdf, jpg and a large variety 
	of other formats. FEN outputs of single trees (ie forest with a single component) are rendered 
	as enewick. Newick files can be visualized in a 	large number of programs 
	(e.g. \href{http://tree.bio.ed.ac.uk/software/figtree/}{FigTree}; 
	\href{http:/https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/}
	{Dendroscope}). 	
	When FEN/Enewick files are output, leaf vertices are modified if they have indegree $>$ 1, 
	creating a new node as parent to that leaf 	and redirecting the leaf's in-edges to that new 
	node with a single edge connecting the new node to the leaf. PDFs of dot files can be 
	generated directly (option \texttt{report}) or manually.  Example dot command line: 
	
	\begin{verbatim}
		dot -Tpdf myDotFile.dot $>$ myDotFile.pdf
	\end{verbatim}
		
	Multiple ``dot'' graphs can be output in a single file. To create pdf and other formats the
	commandline would be (these files are named and numbered automatically):
	
	\begin{verbatim}
		dot -Tpdf -O myDotFile.dot
	\end{verbatim}
		
	For some reason on OSX the `pdf' option does not seem to work. You can use `-Tps2' and 
	that will generate 
	a postscript file ($>$ filename.ps) that Preview can read and convert to pdf.
	%what's -Tps2?	

	

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%COMMANDS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
	
\chapter{PhyG Commands}

\section{\phyg Command Structure}
		
	\subsection{Brief description}
		\phyg interprets and executes scripts coming from an input file. A script is a list of 
		commands, separated by any number of whitespace characters (spaces, tabs, or 
		newlines). Each command consists of a name, followed by an argument or list of 
		arguments separated by commas and enclosed in parentheses. Commands and 
		arguments are case insensitive with the exception of filename specifications, which 
		must match the filename \textit{exactly}, including suffixes and are always in double 
		quotes (\texttt{``fileName''}). All commands, with the exception of \texttt{refine} 
		(Section \ref{subsec:refine}) have default values, which are provided below.
		
		Most arguments are optional, with only a few requiring specification (e.g. the edit 
		operation arguments of \texttt{refine}). Arguments can be followed by options 
		separated by a colon \texttt{(``:'')}, with no space before the option. In this case 
		they are considered labeled-value arguments. There are defaults for all arguments 
		except input graphs. Some Parameters are given with options 
		in a range `a to b' (a-b) with any value in the interval, or alternates `a or b' (a|b). The 
		options parameters associated with 
		
		Wildcard expressions (such as `*' and `?') can be used for input files (data files and 
		graph files (Section \ref{subsec:read}).
			
%		\begin{quote}
%		command(argument, option:argument, option[:optional argument]...)
%		\end{quote}
%		\hl{is this structure above correct?}\\
		
		The following examples illustrate the structure of valid \phyg commands.	
		
		\begin{quote}
		build(distance, replicates:30)
		\end{quote}		
		
		In this first example, the command \texttt{build} is followed by the arguments 
		\texttt{distance} and \texttt{replicates:30}, enclosed in parentheses and separated 
		by a comma. The second argument \texttt{replicates} is a labeled-value argument 
		that contains the label (\texttt{replicates}) and a value (\texttt{30}) separated by a 
		colon.
				
		\begin{quote}
		swap(drift:3, acceptequal:2.0)
		\end{quote}		

		In the second example, the command \texttt{swap} is followed by the arguments 
		\texttt{drift} and \texttt{acceptequal}, enclosed in parentheses and separated 
		by a comma. Both arguments are labeled-value arguments, each with different
		ascribed values. The argument \texttt{drift} is XXX by an integer value (in this case
		\texttt{3}, and the argument \texttt{acceptequal} is XXX by a real number or float
		\texttt{2.0}.
		
	\subsection{Command order and processing}
		The commands \texttt{read}, \texttt{rename}, \texttt{reblock}, and \texttt{set} are executed at
		the beginning of program execution, irrespective of where they appear in the command script. 
		All other commands are executed in the order they are specified. \hl{Command arguments are order independent and can be entered from a file as 
		stdin (\texttt{$<$ fileName})}.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%COMMAND REFERENCE
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section{Command Reference}
	\input{PhyG_Allcommands.tex}

\chapter{Program Usage}
\section{Example Script Files}
	The following file (titled ``Example Script 1'')reads two input sequence files (net-I.fas and net-II.fas), 
	skips all 	the lines that begin with double dash (\texttt{--}), reads the graph file net-I-II.dot, sets the 
	outgroup to the taxon named ``zero,'' specifies the graph type for the analysis is a soft-wired network, 
	and 	reports a series of files with various information about the data and graphs.
	
		\begin{verbatim}
			-- Example Script 1
			read("net-I.fas")
			--read("net-Ia.fas")
			--read("net-IIa.fas")
			read("net-II.fas")
			--read("net-I.dot")
			--read("net-I.tre")
			--read("net-II.tre")
			--read("net-II.dot")
			read("net-I-II.dot")
			set(outgroup:"zero")
			set(graphtype:softwired)
			report("net-test.tre", graphs, newick, overwrite)
			report("net-test.dot", graphs, dot, overwrite)
			report("net-test-data.csv", data, overwrite)
			report("net-test-diag.csv", diagnosis, overwrite)
			report("net-display.dot", displaytrees, dot, overwrite)
			report("net-display.tre", displaytrees, newick, overwrite)
		\end{verbatim}
	
\section{Faster and Slower}
Multiple options affect both the quality of results (better ot worse optimality score) and
the overall execution time.  In general, the more time consuming options also have larger
memory footprints.

\subsection{Evaluation of Graphs}
	\texttt{Multitraverse}\\
	\texttt{CompressResolutions}\\
	\texttt{SoftwiredMethod}\\

\subsection{Search Options}
	\texttt{steepest}\\
	\texttt{Limiting number of network edges}\\
	\texttt{Limiting number of graphs}
	
\section{Parallel Evaluation}
	\texttt{+RTS -NX -RTS}
	
\section{Memory Use}

The amount of memory used during program execution can be reported by adding the 
runtime option ``-s'' as in \texttt{+RTS -s -RTS} to the command line (runtime options can 
be specified together as in \texttt{+RTS -NX -s -RTS}). This will output several fields of 
data with the  ``in use'' field specifying the maximum amount of memory requested from 
the OS. RTS options are described in detail at \url{https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html}. 
	
\chapter*{Acknowledgments}
	The authors would like to thank DARPA SIMPLEX N66001-15-C-4039, the Robert J. 
	Kleberg Jr. and Helen C. Kleberg foundation grant ``Mechanistic Analyses of Pancreatic 
	Cancer Evolution'', and the American Museum of Natural History for financial support. 
	
	%\newpage
	%\bibliography{big-refs-3.bib}
	\bibliography{/Users/louise/DropboxAMNH/big-refs-3.bib}
	%\bibliography{/home/ward/Dropbox/Work_stuff/manus/big-refs-3.bib}
	%\bibliography{/users/ward/Dropbox/work_stuff/manus/big-refs-3.bib}
 \end{document}