\documentclass[11pt]{book}
\usepackage{longtable}
\usepackage{color}
\usepackage{tabu}
\usepackage{setspace}
\usepackage{pdflscape} 
\usepackage{graphicx}
\usepackage {float}
%\usepackage{subfigure}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{natbib}
\usepackage{fullpage}
\bibliographystyle{plain}
%\bibliographystyle{cbe}
\usepackage{algorithmic}
\usepackage[vlined,ruled]{algorithm2e}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage[T1]{fontenc}
\usepackage{url}
\usepackage[dvipsnames]{xcolor}
\usepackage{color, soul}
\usepackage[colorlinks=true, linkcolor=blue, citecolor=DarkOrchid, urlcolor=TealBlue ]{hyperref}
%\usepackage[nottoc,numbib]{tocbibind}
\usepackage{tocloft}
\usepackage[many]{tcolorbox}
\usepackage{marginnote}
\usepackage{lipsum}
%\usepackage[most]{tcolorbox}

\setlength\itemindent{1cm}

\newcommand{\phyg}{\texttt{PhyG} }
\newcommand{\atsymbol}{@}
\newenvironment{phygdescription}{\subsubsection{Description}}{}
\newenvironment{example}{\subsubsection{Examples} \begin{itemize}}{\end{itemize}}
\newenvironment{argument}{\subsection{Arguments}\begin{itemize}}{\end{itemize}}
%\item [ ]

\begin{document}
	%\firstpage{1}
	
	\title{PhylogeneticGraph\\User Manual\\Version 0.1}
	
	\maketitle
	
	\newpage

	 \begin{center}
		\includegraphics[width=\textwidth]{AMNHLogo.pdf}
	\end{center}

	\vspace*{2.50cm}	
	\begin{flushleft}
		\textbf {Program and Documentation} \\ Ward C. Wheeler \\
		\vspace*{0.50cm}
		\textbf {Program} \\ Alex Washburn \\
		\vspace*{0.50cm}
		\textbf{Documentation} \\ Louise M. Crowley
	\end{flushleft}
	

	\vspace*{2.50cm}
	
	\begin{flushleft}
		\small
		{\it Louise M. Crowley, Alex Washburn, Ward C. Wheeler} \\
		
		Division of Invertebrate Zoology, American Museum of Natural History, New York, NY, U.S.A.\\
		\smallskip
		The American Museum of Natural History\\
		\copyright 2023 by The American Museum of Natural History, \\
		All rights reserved. Published 2023.
		
		%	\vspace*{0.25cm}
		%	\emph{W. C. Wheeler.} 2022. \texttt{PHYG} 1..0. New York, 
		%	American Museum of Natural History. Documentation by W.C. Wheeler and L. M. Crowley. 
		%	
		\vspace*{0.25cm}
		
		Available online at \url{https://github.com/amnh/PhyGraph}
		
		Comments or queries relating to the documentation should be sent to \href{mailto:wheeler@amnh.org}
		{wheeler@amnh.org} or \href{mailto:crowley@amnh.org}{crowley@amnh.org}
	\end{flushleft}
	
	\tableofcontents

\chapter{What is PhyG?}

\section{Introduction}
	PhylogeneticGraph (\texttt{PhyG}) is a multi-platform program designed to produce phylogenetic 
	graphs from input data and graphs via heuristic searching of general phylogenetic graph 
	space. \texttt{PhyG} is the successor of \href{https://github.com/wardwheeler/POY5}{\textbf{POY}}
	\citep{POY2,POY3,POY4,Varonetal2010,POY5, Wheeleretal2015}, containing much of its 
	functionality, including the optimization of \textit{unaligned} sequences, and the ability to implement 
	search strategies such as random addition sequence, swapping, and tree fusing. As in {\textbf{POY}, 
	\phyg can generate outputs in the form of implied alignments and graphical representations of 
	cladograms and graphs. What sets \phyg apart from {\textbf{POY}, and other phylogenetic 
	analysis programs, is the extension to broader classes of input data and phylogenetic graphs. 
	The phylogenetic graph inputs and outputs of \texttt{PhyG} include trees, as well as softwired 
	and hardwired networks.
		
	This is the initial version of documentation for the program.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%QUICKSTART
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%		
\section{Quick Start}
	
	\subsection{Requirements: software and hardware}
		\phyg is an open-source program that can be compiled for Mac OSX and Linux 
		(see information relating to Windows machines below). Some utility programs 
		(such as TextEdit for Mac, or Nano for Linux) can help in preparing \phyg scripts 
		and formatting data files, while others (such as Adobe Acrobat and TreeView 
		\citep{page1996}) can facilitate viewing the outputted graphs and trees.
		
		\phyg runs on a variety of computers, including desktops, laptops and cluster computers.
		By default, \phyg is a multi-threaded application and will use the available resources of 
		the computer during the analysis (see Execution in Parallel \ref{subsec:parallel}). 
		
	\subsection{Obtaining and Installing \phyg}
		\phyg source code, precompiled binaries, test data, and documentation in pdf format, 
		as well as tutorials, are available from the \phyg \href{https://github.com/amnh/PhyGraph}{GitHub} 
		website.

	\subsubsection{Installing from the binaries}
		Download the \phyg binary from the \href{https://github.com/amnh/PhyGraph}{GitHub} 
		website. Binaries are available for Mac OSX computers with either Intel or M1 processors, 
		and Linux machines (see information relating to Windows machines below).
		The user can go directly to the website and click on the appropriate link 
		for the binary. On most systems this will download to either your Desktop or Downloads folder. \\
		
		Alternatively, open a \textit{Terminal} window (located in your Applications folder) and type 
		the following for either the Mac Intel, Mac M1 or Linux binary:
		
		\begin {quote}
		curl -LJ --output phyg https://github.com/amnh/PhyGraph/blob/main/bin/OSX/phyg-Intel?raw=true
		\end{quote}		
		
		\noindent or 
		
		\begin {quote}
		curl -LJ --output phyg https://github.com/amnh/PhyGraph/blob/main/bin/OSX/phyg-M1?raw=true
		\end{quote}	
		
		\noindent or 
		
		\begin {quote}
		curl -LJ --output phyg https://github.com/amnh/PhyGraph/blob/main/bin/linux/phyg?
		raw=true
		\end{quote}
		
		\noindent The binary should either be moved into your \$PATH or referred to its 
		absolute when executing a script.\\
		
		For those users with Windows machines, a Windows Subsystem for Linux 
		(WSL) can be installed. This system will allow you to run the Linux binary directly 
		on your machine, without having to install a virtual machine or dual-boot setup. 
		The WSL, along with directions for installation, can be found 
		\href{https://learn.microsoft.com/en-us/windows/wsl/}{here}.
	
	\subsubsection{Compiling from the source}
		For the majority of users, downloading the binaries will suffice. Should the user prefer to 
		compile \phyg directly from the source, the source code can be downloaded 
		from the \href{https://github.com/amnh/PhyGraph}{GitHub} website. \phyg is largely 
		written in Haskell. In order to compile \phyg from the source, the user must install Cabal, 
		a command-line program for downloading and building software written in Haskell (GHC).  
		Information on its installation can be found  
		\href{https://www.schoolofhaskell.com/user/simonmichael/how-to-cabal-install}{here}.
		To compile an optimized version of \texttt{PhyG}, open a \textit{Terminal} and run the 
		following:
		
		\begin{quote}
		%cabal install PhyGraph:phyg --project-file=cfg/cabal.project.release
		cabal build PhyGraph:phyg --flags=super-optimization
		\end{quote}
		
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%OVERVIEW OF THE PROGRAM
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	 		
\section{Overview of program use}
	At present, \phyg is operated solely via command-line in a \textit{Terminal} window
	and cannot be executed interactively. Commands are entered via a script file 
	containing commands that specify input files, output files and formats, graph type 
	and search parameters.
		
	\subsection{Executing Scripts}
		The program is invoked from the command-line as in:
		
		\begin{quote}
		phyg commandFile
		\end{quote}
		
		\smallskip
		
		\noindent For example, typing the following in a \textit{Terminal} window will invoke 
		\phyg to run the script \texttt{mol.pg}, which is located in the Desktop folder 
		\texttt{phygfiles}:
		
		\begin{quote}
  		phyg /Users/Ward/Desktop/phygfiles/mol.pg
		\end{quote}
		
		\bigskip
		
		\noindent 
		This is the equivalent of typing the following from any location on your computer:
		
		\begin{quote}
   		cd ("/Users/Ward/Desktop/phygfiles")
		\end{quote}
			
		\subsection{Creating and running \phyg scripts}
		A script is a simple text file containing a list of commands to be performed. 
		This script can also include references to other script files 
		(Figure \ref{firstscript}).
		
		Scripts can be created using any conventional text editor such as \textit{TextEdit}, 
		\textit{TextWrangler}, \textit{BBEdit}, or \textit{Nano}. Do not use a word processing 
		application like \textit{Microsoft Word} or \textit{Apple Pages}, as these programs 
		can introduce hidden characters in the file, which will be interpreted by \phyg and 
		can cause unpredictable parsing of the data. Comments that describe the contents 
		of the file and provide other useful annotations can be included. Comment lines are 
		prepended with `-{}-' and can span multiple lines, provided each line begins with `-{}-'. 

%Comemnt

		\begin{figure}[H]
		\centering
		\includegraphics[width=\textwidth]{First_run.jpg}
		\caption{\phyg scripts. The headers in the scripts are comments, leading with `-{}-', 
		which is ignored by \phyg. This first script ``\textbf{First\_script.pg}'' includes a reference 
		to the second script ``\textbf{Chel\_files.txt}'', which includes a group of data files to be 
		read by the program.}
		\label{firstscript}
		\end{figure}
	
	\subsection{Execution in Parallel}
	\label{subsec:parallel}
		\phyg is a multi-threading application and will, by default,  use
		all available cores. Should the user wish to limit or 
		specify the number of processors used by \phyg this can be achieved by including the 
		options `\textbf{+RTS -NX -RTS}', where `\textbf{X}' is the number of processors offered 
		to the program, when executing the script. Should the user wish to use a single processor, 
		this can be specified by typing:

		\begin{quote}
		phyg fileName +RTS -N1 
		\end{quote}
		
		\medskip
		\noindent This will execute the program sequentially.		
		%possible to include and other options

		
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%FORMATS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section{Input Data Formats} 
	\phyg can analyze a number of different data types, including qualitative, nucleotide
	and other sequences  
	(aligned and unaligned), in TNT, fasta, and fastc formats. Any character names in 
	input files are (for now) ignored and internal names are created by appending 
	the character number in its file to the filename as in ``\textbf{fileName:0}''. 
	Qualitative data, and prealigned data include their indices within input files and 
	unaligned data are treated as single characters.
	
	\phyg allows the user to comment out portions of a taxon name 
	in the imported data file. This is achieved by inserting a dollar sign (`\$') before the region 
	of text that the user wishes to comment out. As an example, placing a `\$' before the 
	GenBank information in the taxon name \textbf{Lingula\_anatina\$\_AB178773\_16S} 
	will comment out this information and the taxon name will be read as 
	\textbf{Lingula\_anatina} by the program. This can be useful for housekeeping purposes, 
	when it is desirable to maintain long verbose taxon names (such as catalog or NCBI 
	accession numbers) associated with the original data files but avoid reporting these 
	names on the graphs. Moreover, it allows the user to provide a single name for a terminal
	in cases where the corresponding data are stored in different files under different terminal
	names.
		
	\subsection{fasta}
		Single character sequence input \citep{PearsonandLipman1988} (see Figure 
		\ref{fasta-c}).
		
	\subsection{fastc}
		Fastc is a file format for multi-character sequence input \citep{WheelerandWashburn2019}.
		This format is derived from the fasta format and the custom alphabet format of 
		 \href{https://github.com/wardwheeler/POY5}{\textbf{POY}} \citep{POY4,POY5}. 
		 Multi-character sequence elements are useful for the analysis of data types such 
		 as developmental, gene synteny, and comparative linguistic data (see Figure 
		 \ref{fasta-c}).  In this format, individual sequence elements are separated by a space.
		 
		\begin{figure}[H]
		\centering
		\includegraphics[width=0.48\textwidth]{fasta.png}
		\includegraphics[width=0.48\textwidth]{fastc.png}
		\caption{Fasta and Fastc file formats. The file ``\textbf{chel\_cox1aln1.fasta}'' 
		represents a fasta file with a greater than sign (`>') preceding taxon names and nucleotide  
		sequence data following on a new line. The file ``\textbf{woman.fastc}'' is a fastc 
		file, with a greater than sign (`>') preceding taxon names and the linguistic data following 
		on a new line.}
		\label{fasta-c}
		\end{figure}
		
	\subsection{\texttt{TNT}}
	\label{subsec:TNT}
		The TNT \citep{Goloboffetal2008} format is accepted here for specification of 
		qualitative, measurement, and prealigned molecular sequence data. \phyg can 
		parse all of the basic character settings including activation/deactivation of 
		characters, making characters additive/non-additive, applying weight to characters, 
		step-matrix costs, interleaved data files and continuous characters (see the 
		argument \texttt{tnt} in \ref{subsec:read}). Continuous characters can only be 
		in the format of integers or floats. Moreover, they must be declared as ``additive'', 
		otherwise, they will be treated as non-additive character states (see 
		\href{http://phylo.wikidot.com/tnt-htm}{\textbf{TNT}}).
		The characters `-' and `?' can be used to indicate that characters or character 
		states are missing or inapplicable. Only one set of ccode commands is allowed 
		per line (e.g. to make everything additive: ccode +.;). Default values of step-matrix 
		weights are not implemented, all step-matrix values must be individually specified. 
		Ambiguities or ranges are denoted in square brackets (`\texttt{[ ]}') with a period, 
		as in [X.Y]. Continuous characters are denoted in square brackets with a dash or 
		hyphen, as in [X-Y].

\section{Input Graph Formats}
	Graphs can be input in a number of formats, including newick, enhanced newick and 
	dot. Newick tree file format is the parenthetical representation of trees as interpreted 
	by Olsen (linked \href{https://evolution.genetics.washington.edu/phylip/newick_doc.html}
	{here}). Enhanced Newick \cite{Cardonaetal2008} is an extension of Newick file 
	format, with the addition of tags (`\#') that indicate hybrid zones or reticulation events
	in the phylogenetic network. \href{https://graphviz.org/}{Dot} is a graph description 
	language and well suited to represent graphs and networks. 	
	%and Forest Enhanced Newick (as defined by Wheeler, 2022 \cite{Wheeler2022}). 
	%Forest Enhanced Newick (FEN) is a format based on Enhanced Newick (ENewick) for 
	%forests of components, each of which is represented by an ENewick string. The ENewick 
	%components are surrounded by `$<$' and '$>$'. As in $<$(A, (B,C)); (D,(E,F));$>$. 
	%Groups may be shared among ENewick components.
	
	\begin{figure}%[H]
	\centering
	\includegraphics[width=\textwidth]{enewick.png}
	\caption{The file \textbf{``flu\_net1.tre''} in Enhanced Newick (ENewick) graph format. The 
	values associated with the taxon names and HTUs are branch lengths. The cost of the 
	graph(s) can be found in square brackets at the end of each graph. A semi-colon follows
	the cost of each graph as a comment.}
	\label{enewick}
	\end{figure}
	
\section{Output Graph Formats}
\label{sec:outputgraphs}
	Graph outputs can be in either newick, enewick, dot, and depending on the 
	operating system, eps or pdf formats. Newick tree files can be viewed in other 
	programs like \href{http://tree.bio.ed.ac.uk/software/figtree/}{FigTree} or 	
	\href{http:/https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/}{Dendroscope}. 
	Enewick files can only be viewed in a text editor. \phyg can output a dot file, 
	along with an eps (on OSX) or pdf (on linux) file that can be viewed in a vector 
	graphics program. The dot file can be viewed (and modified) in \textit{Graphviz}. 
	Note: in order to output pdf files the application \textit{dot} must be installed from 
	the \href{https://graphviz.org/download/}{Graphviz} website. Graphviz an open-source
	graph visualization software. \\	
	
	\noindent PDFs can also be generated directly from dot files. In a \textit{Terminal}, 
	type the following: 

	\begin{quote}
	dot -Tpdf myDotFile.dot $>$ myDotFile.pdf
	\end{quote}
		
	\noindent Multiple `dot' graphs can be output in a single file. To create pdf and 
	other formats the commandline would be (these files are named and numbered 
	automatically):
	
	\begin{quote}
	dot -Tpdf -O myDotFile.dot
	\end{quote}
		
	\noindent In OSX the `pdf' option does not currently seem to work. In this case, 
	type the following:

	\begin{quote}
	dot -Tps2 myDotFile.dot > myDotFile.ps
	\end{quote}
	
	\noindent `-Tps2' will generate a postscript file that \textit{Preview} can read and 
	convert to pdf.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%COMMANDS
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
	
\chapter{PhyG Commands}

\section{\phyg Command Structure}
		 
	\subsection{Brief description}
		\phyg interprets and executes scripts coming from an input file. A script is a list of 
		commands, separated by any number of whitespace characters (spaces, tabs, or 
		newlines). Each command consists of a name, followed by an argument or list of 
		arguments separated by commas (`\texttt{,}') and enclosed in parentheses. 
		Commands and arguments are case insensitive with the exception of filename 
		specifications, which must match the filename \textit{exactly}, including suffixes 
		and are always in double quotes (``\textbf{fileName}''). Most commands, with the 
		exception of \texttt{reblock}, \texttt{rename} and \texttt{transform}, have default values 
		and are provided at the end of each command section below. If no arguments are 
		specified, the command is executed under its default settings.\\
		
		\noindent \phyg recognizes three types of command arguments: \textit{primitive values}, 
		\textit{labeled values}, and \textit{lists of arguments}.\\
		
		\noindent \textbf{Primitive values} can either be an integer (\texttt{INT}), 
		a real number (\texttt{FLOAT}), a string (\texttt{STRING}) or a boolean (\texttt{BOOL)} 
		i.e. True|False. \\
		
		\noindent \textbf{Labeled arguments} consist of an argument and an identifier 
		(the label), separated by the colon (`\texttt{:}') character. Examples of 
		identifiers include \texttt{majority}, \texttt{nopenalty}, \texttt{hardwired}, and 
		\texttt{parsimony}. \\
		
		\noindent \textbf{List of arguments} is several arguments enclosed in parenthesis 
		and separated by commas (`\texttt{,}'). Most arguments are optional, with only a 
		few requiring specification, e.g. the build method of distance must be specified. 
		%Some argument labels are obligatory, most are optional \hl{e.g. ?}. 
		In cases where an argument must be specified, \phyg will report a warning message 
		in the output of the terminal window. When no arguments are specified, \phyg will 
		use the default values.\\
		
		\noindent The following examples illustrate the structure of valid \phyg commands. 
					
		\begin{quote}
		build()
		\end{quote}
		
		\noindent In this simple example, the command \texttt{build} is followed by an open 
		and closed parenthesis. As no arguments are specified, \phyg will use the defaults, 
		so this is equivalent to \texttt{build(character, replicates:10)}.	
			
		\begin{quote}
		read(nucleotide:"chel-prealigned.fas", tcm:"sg1t4.mat")
		\end{quote}		

		\noindent In this second example, the command \texttt{read} reads data in the file
		(\texttt{STRING}) ``\textbf{chel-prealigned.fas}'', parsing it as nucleotide sequence data. 
		It also read the data in the file ``\textbf{sg1t4.mat}'', parsing it as a transformation cost matrix, 
		with indel and substitution costs set to $1$.
		
		\begin{quote}
		swap(drift:3, acceptequal:2.0)
		\end{quote}
				
		\noindent In the third example, the command \texttt{swap} is followed by the list 
		of arguments that includes \texttt{drift} and \texttt{acceptequal}, enclosed in 
		parentheses and separated by a comma. Both of these are labeled-value arguments, 
		ascribed an \texttt{INTEGER} and a \texttt{FLOAT}, respectively.
		
		\begin{quote}
		set(graphfactor:nopenalty) 
		\end{quote}
		
		\noindent In this fourth and final example, the command \texttt{set} is followed by the 
		labeled-value argument \texttt{graphfactor} with the label \texttt{nopenalty}.
				
	\subsection{Command order and processing}
		The commands \texttt{read}, \texttt{rename}, \texttt{reblock}, and \texttt{set} are 
		executed at the beginning of program execution, irrespective of where they appear 
		in the command script. All other commands are executed in the order they are specified. 
		
	\subsection{Notation}
		Commands are listed alphabetically in the next section. Commands are shown in 
		\texttt{terminal} typeface. Optional items are enclosed in square brackets (`\texttt{[ ]}').
		Primitive values are shown in \texttt{UPPERCASE}. 
	

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%COMMAND REFERENCE
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section{Command Reference}
	\input{PhyG_Allcommands.tex}

%\chapter{Program Usage}
%\section{Example Script Files}
%Tutorials are available for download on the \phyg \href{https://github.com/amnh/PhyGraph}{GitHub}
%website. These tutorials provide guidance for using the program. Each tutorial contains a \phyg script 
%that includes detailed commentaries explaining the rationale behind each step of the analysis. The 
%command arguments will differ substantially depending on type, complexity, and size of the data set. 
%
%%	The following file (titled ``Example Script 1'') reads two input sequence files (net-I.fas and net-II.fas), 
%%	skips all 	the lines that begin with double dash (\texttt{--}), reads the graph file net-I-II.dot, sets the 
%%	outgroup to the taxon named ``zero,'' specifies the graph type for the analysis is a softwired network, 
%%	and 	reports a series of files with various information about the data and graphs.
%%	
%%		\begin{verbatim}
%%			-- Example Script 1
%%			read("net-I.fas")
%%			--read("net-Ia.fas")
%%			--read("net-IIa.fas")
%%			read("net-II.fas")
%%			--read("net-I.dot")
%%			--read("net-I.tre")
%%			--read("net-II.tre")
%%			--read("net-II.dot")
%%			read("net-I-II.dot")
%%			set(outgroup:"zero")
%%			set(graphtype:softwired)
%%			report("net-test.tre", graphs, newick, overwrite)
%%			report("net-test.dot", graphs, dot, overwrite)
%%			report("net-test-data.csv", data, overwrite)
%%			report("net-test-diag.csv", diagnosis, overwrite)
%%			report("net-display.dot", displaytrees, dot, overwrite)
%%			report("net-display.tre", displaytrees, newick, overwrite)
%%		\end{verbatim}
%	
%\section{Faster and Slower}
%Multiple options affect both the quality of results (better ot worse optimality score) and
%the overall execution time.  In general, the more time consuming options also have larger
%memory footprints.
%
%\subsection{Evaluation of Graphs}
%	\texttt{Multitraverse}\\
%	\texttt{CompressResolutions}\\
%	\texttt{SoftwiredMethod}\\
%
%\subsection{Search Options}
%	\texttt{steepest}\\
%	\texttt{Limiting number of network edges}\\
%	\texttt{Limiting number of graphs}
%	
%\section{Parallel Evaluation}
%	\texttt{+RTS -NX -RTS}
%	
%\section{Memory Use}
%
%The amount of memory used during program execution can be reported by adding the 
%runtime option ``-s'' as in \texttt{+RTS -s -RTS} to the command line (runtime options can 
%be specified together as in \texttt{+RTS -NX -s -RTS}). This will output several fields of 
%data with the  ``in use'' field specifying the maximum amount of memory requested from 
%the OS. RTS options are described in detail at \url{https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html}. 
%	
	% Optimize memory consumption--keep low number of graphs in initial searches, later keep 
% a larger number to get others 
%		\hl{(see email from WW 04-21-22)}
%		\hl{where should I discuss this?}


\section*{Acknowledgments}
	The authors would like to thank DARPA SIMPLEX N66001-15-C-4039, the Robert J. 
	Kleberg Jr. and Helen C. Kleberg foundation grant ``Mechanistic Analyses of Pancreatic 
	Cancer Evolution'', and the American Museum of Natural History for financial support. 
	
	%\newpage
	%\bibliography{big-refs-3.bib}
	\bibliography{/Users/louise/DropboxAMNH/big-refs-3.bib}
	%\bibliography{/home/ward/Dropbox/Work_stuff/manus/big-refs-3.bib}
	%\bibliography{/users/ward/Dropbox/work_stuff/manus/big-refs-3.bib}
 \end{document}